Kubernetes faq etcd quota (#1225)
ovh/docs: pages/platform/kubernetes-k8s/etcd-quota-error/guide.en-gb.md

> ---

Reviewer: Don't forget to create the other language files when the review has been fully applied

Author: yep i was waiting the entire doc is review and accepted

title: ETCD Quotas error, troubleshooting
excerpt: ''
slug: etcd-quota-error
section: FAQ
---

**Last updated March 15<sup>th</sup>, 2021.**

> A lot of customers encounter some trouble with the quota on ETCD allowed disk space.

Reviewer: ```suggestion
Some customers might encounter troubles with the quota on ETCD storage usage.
```

> ## Issues

Reviewer: ```suggestion
## Issue
```

> When you want to manage an objects into your Kubernetes cluster, you can reached an error like this:

Reviewer: ```suggestion
When managing your Kubernetes cluster, you may encounter the following error:
```

```log
"Error from server: rpc error: code = Unknown desc = ETCD storage quota exceeded"
```

> Each Kubernetes cluster have a dedicated quota on ETCD storage.
> 
> The formula to calculate ETCD quota allowed is: *Total = 10MB + (25MB per node)*

Reviewer: ```suggestion
Each Kubernetes cluster has a dedicated quota on ETCD storage usage,
calculated through the following formula:
*Quota = 10MB + (25MB per node)*
```

Reviewer: EDIT: We will soon add a max of 200MB to the formula

```suggestion
Each Kubernetes cluster has a dedicated quota on ETCD storage usage,
calculated through the following formula:
*Quota = 10MB + (25MB per node)* (capped to 200MB)
```

> For a cluster with 3 servers "b2-7" the quota are : 85MB

Reviewer: ```suggestion
For example, a cluster with 3 `b2-7` servers has a quota of 85MB.
```

> Don't worry about the decrease, you keep the max quotas allowed forever, to prevent shrink of datas.

Reviewer: ```suggestion
The quota can thus be increased by adding nodes, but will never be decreased (even if all nodes are removed) to prevent data loss.
```

> ## Troubleshoot

Reviewer: ```suggestion
## Resolution

The error mentioned above states that the cluster's ETCD storage usage has exceeded the quota.
To resolve the situation, resources created in excess need to be deleted.
```

### Cert-Manager

> The first mostfull usecase is come from with a bad configuration on [cert-manager](https://cert-manager.io/docs/)
> 
> Most users install the cert-manager from Helm, and take a quick road to prod it.
> 
> If cert-manager have a bad configuration, he try to create unlimited number of certificare requets.
> 
> The operator cert-manager going to satured all empty allow storage in ETCD.

Reviewer: ```suggestion
Most users install cert-manager through Helm, and then move on a bit hastily.

The most common cases of ETCD quota issues come from a bad configuration of [cert-manager](https://cert-manager.io/docs/),
making it continuously create `certificaterequest` resources.

This behaviour will fill the ETCD with resources until the quota is reached.
```

> To detect that, you can find the number of certificate-requets:

Reviewer: ```suggestion
To detect this situation, you can get the number of `certificaterequest` resources:
```

```bash
kubectl get certificaterequests -A | wc -l
```

> If you have a huge number of certificate request (+1000 for example), you have find the root cause !

Reviewer: ```suggestion
If you have a huge number of certificate requests (+1000 for example), you have found the root cause.
```

> To solve that,we propose this way:

Reviewer: ```suggestion
To resolve the situation, we propose the following way:
```

> * Stop the operator cert-manager

Reviewer: ```suggestion
* Stop cert-manager
```

```bash
kubectl scale deployment --replicas 0 cert-manager
```

> * flush all useless certificate request

Reviewer: ```suggestion
* Flush all pending certificaterequests
```

```bash

> kubectl get 
> 
> // TODO a fixer
> 

Reviewer: ```suggestion
```

kubectl get certificates --all-namespaces -o json \
    | jq -r '.items[] | select(.status.phase == "Pending").metadata.namespace' \
    | parallel kubectl -n '{}' delete po -l 'component=kube-controller-manager'
```

* Update cert-manager

> In this case, we don't have a generic solution, but if you have use Helm, we recommend to use Helm for the update.

Reviewer: ```suggestion
There is no generic way to do this, but if you use Helm we recommend you to use it for the update.
```

[Cert Manager official documentation](https://cert-manager.io/docs/installation/kubernetes/)

* Fix the issue

> We recommend you to follow the steps for troubleshoot, and ensure all is correctly configured

Reviewer: ```suggestion
We recommend you to take the following steps to troubleshoot your cert-manager, and to ensure that everything is correctly configured:
```

[Acme troubleshoot](https://cert-manager.io/docs/faq/acme/)

> * Start the cert-manager operator

Reviewer: ```suggestion
* Start cert-manager
```

### Other usecase

> If cert-manager is not the root causes, you should turn to the other installed operator.

Reviewer: ```suggestion
If cert-manager is not the root causes, you should turn to the other running operators.
```

> The way is to find the objects creates by the operator, it is possible the operator create a lots of objects like cert-manager.

Reviewer: ```suggestion
We have found that the following resources can sometimes be generated continuously by existing operators:
> backups.velero.io
ingress.networking.k8s.io
ingress.extensions

If that still  does not cover your case, you can use a tool like [ketall](https://github.com/corneliusweig/ketall) to easily list and count resources in your cluster.
```
